File: backend/internal/idls/chart.go
package idls

type ChartType string

const (
	ChartTypeLine    ChartType = "line"
	ChartTypeBar     ChartType = "bar"
	ChartTypePie     ChartType = "pie"
	ChartTypeScatter ChartType = "scatter"
	ChartTypeTable   ChartType = "table"
)

type CreateChartRequest struct {
	Name      string    `json:"name" binding:"required"`
	DatasetID int       `json:"dataset_id" binding:"required"`
	ChartType ChartType `json:"chart_type" binding:"required"`
	Config    string    `json:"config" binding:"required"`
}

type UpdateChartRequest struct {
	Name      *string    `json:"name,omitempty"`
	DatasetID *int       `json:"dataset_id,omitempty"`
	ChartType *ChartType `json:"chart_type,omitempty"`
	Config    *string    `json:"config,omitempty"`
}

type ChartResponse struct {
	ID        int    `json:"id"`
	Name      string `json:"name"`
	DatasetID int    `json:"dataset_id"`
	ChartType string `json:"chart_type"`
	Config    string `json:"config"`
	CreatedAt string `json:"created_at,omitempty"`
	UpdatedAt string `json:"updated_at,omitempty"`
}

type ChartListResponse struct {
	Items []ChartResponse `json:"items"`
	Total int64           `json:"total"`
}

type ChartQueryAggregation string

const (
	ChartQueryAggSum   ChartQueryAggregation = "sum"
	ChartQueryAggAvg   ChartQueryAggregation = "avg"
	ChartQueryAggCount ChartQueryAggregation = "count"
	ChartQueryAggMax   ChartQueryAggregation = "max"
	ChartQueryAggMin   ChartQueryAggregation = "min"
)

type ChartQueryMetric struct {
	Field string                `json:"field" binding:"required"`
	Agg   ChartQueryAggregation `json:"agg" binding:"required"`
	Alias *string               `json:"alias,omitempty"`
}

type ChartQueryFilterOp string

const (
	FilterOpEq        ChartQueryFilterOp = "eq"
	FilterOpNeq       ChartQueryFilterOp = "neq"
	FilterOpGt        ChartQueryFilterOp = "gt"
	FilterOpGte       ChartQueryFilterOp = "gte"
	FilterOpLt        ChartQueryFilterOp = "lt"
	FilterOpLte       ChartQueryFilterOp = "lte"
	FilterOpLike      ChartQueryFilterOp = "like"
	FilterOpIn        ChartQueryFilterOp = "in"
	FilterOpBetween   ChartQueryFilterOp = "between"
	FilterOpIsNull    ChartQueryFilterOp = "isNull"
	FilterOpIsNotNull ChartQueryFilterOp = "isNotNull"
)

type ChartQueryFilter struct {
	Field    string             `json:"field" binding:"required"`
	Op       ChartQueryFilterOp `json:"op" binding:"required"`
	Value    any                `json:"value"`
	ValueEnd *any               `json:"value_end,omitempty"`
	Logic    string             `json:"logic"`
}

type ChartQueryPagination struct {
	Page     int `json:"page"`
	PageSize int `json:"page_size"`
}

type ChartQuerySort struct {
	Field string `json:"field" binding:"required"`
	Order string `json:"order" binding:"required"`
}

type ChartQueryRequest struct {
	DatasetID  int                   `json:"dataset_id" binding:"required"`
	ChartType  string                `json:"chart_type" binding:"required"`
	Dims       []string              `json:"dims"`
	Metrics    []ChartQueryMetric    `json:"metrics"`
	Filters    []ChartQueryFilter    `json:"filters"`
	Pagination *ChartQueryPagination `json:"pagination,omitempty"`
	Sort       *ChartQuerySort       `json:"sort,omitempty"`
}

type TableResponse struct {
	Columns    []string         `json:"columns"`
	Data       []map[string]any `json:"data"`
	Pagination *TablePagination `json:"pagination"`
}

type TablePagination struct {
	Page       int   `json:"page"`
	PageSize   int   `json:"page_size"`
	Total      int64 `json:"total"`
	TotalPages int   `json:"total_pages"`
}

type PieResponseItem struct {
	Name       string  `json:"name"`
	Value      float64 `json:"value"`
	Percentage float64 `json:"percentage"`
}

type PieResponse struct {
	Data []PieResponseItem `json:"data"`
}

type AxisSeries struct {
	Name string `json:"name"`
	Data any    `json:"data"`
}

type AxisResponse struct {
	XAxis  []string     `json:"x_axis"`
	Series []AxisSeries `json:"series"`
}

type ScatterResponse struct {
	Data [][2]float64 `json:"data"`
}

type ChartDataResponse = any

File: frontend/src/idls/chart.ts
export type ChartType = 'line' | 'bar' | 'pie' | 'scatter' | 'table';
export type ChartQueryAggregation = 'sum' | 'avg' | 'count' | 'max' | 'min';

export interface CreateChartRequest {
  name: string;
  dataset_id: number;
  chart_type: ChartType;
  config: string;
}

export interface UpdateChartRequest {
  name?: string;
  dataset_id?: number;
  chart_type?: ChartType;
  config?: string;
}

export interface ChartResponse {
  id: number;
  name: string;
  dataset_id: number;
  chart_type: string;
  config: string;
  created_at?: string;
  updated_at?: string;
}

export interface ChartListResponse {
  items: ChartResponse[];
  total: number;
}

export interface ChartQueryMetric {
  field: string;
  agg: ChartQueryAggregation;
  alias?: string;
}

export type ChartQueryFilterOp = 'eq' | 'neq' | 'gt' | 'gte' | 'lt' | 'lte' | 'like' | 'in' | 'between' | 'isNull' | 'isNotNull';

export interface ChartQueryFilter {
  field: string;
  op: ChartQueryFilterOp;
  value: unknown;
  value_end?: unknown;
  logic: 'and' | 'or';
}

export interface ChartQueryPagination {
  page: number;
  page_size: number;
}

export interface ChartQuerySort {
  field: string;
  order: 'asc' | 'desc';
}

export interface ChartQueryRequest {
  dataset_id: number;
  chart_type: string;
  dims: string[];
  metrics: ChartQueryMetric[];
  filters: ChartQueryFilter[];
  pagination?: ChartQueryPagination;
  sort?: ChartQuerySort;
}

export interface TableResponse {
  columns: string[];
  data: Record<string, unknown>[];
  pagination: {
    page: number;
    page_size: number;
    total: number;
    total_pages: number;
  };
}

export interface PieResponseItem {
  name: string;
  value: number;
  percentage: number;
}

export interface PieResponse {
  data: PieResponseItem[];
}

export interface AxisSeries {
  name: string;
  data: unknown[];
}

export interface AxisResponse {
  x_axis: string[];
  series: AxisSeries[];
}

export interface ScatterResponse {
  data: [number, number][];
}

export interface GeneratedSQL {
  select: string;
  count: string;
}

export interface ChartQueryResponseData {
  data: ChartDataResponse;
  generated_sql?: GeneratedSQL;
}
